fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,680 @@
+from fairscale.nn.pipe.checkpoint import Checkpointing, TensorOrTensors
+from fairscale.nn.pipe.dependency import fork, join
+from fairscale.nn.pipe.microbatch import Batch
+from fairscale.nn.pipe.stream import as_cuda, current_stream, is_cuda, use_device, use_stream
+from fairscale.nn.pipe.worker import Task, create_workers
 
+Device = Union[torch.device, int, str]
 
-def _rcat(tensors: List) -> Tensor:
-    return torch.cat([t.local_value() for t in tensors])
+ExcInfo = Tuple[Type[BaseException], BaseException, TracebackType]
 
 
-def _rcheckpoint(rmodule: rpc.RRef, input_rref: rpc.RRef) -> TensorOrTensors:
-    module = rmodule.local_value()
-    input = module[0](input_rref)  # calls _ToHere.forward
-    return checkpoint_sequential(module[1:], 1, input)
+def check_pytorch_version() -> None:
+    if torch.__version__.split("+")[0].split(".")[:2] < ["1", "9"]:
+        raise Exception("DistributedPipeline requires PyTorch version 1.9 or higher")
 
 
-def _parameter_rrefs(module: rpc.RRef) -> List[rpc.RRef]:
-    return [rpc.RRef(p) for p in module.local_value().parameters()]
-
-
-def _rloss(loss_func: Callable, input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+def rloss(loss_func: Callable, input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
     return loss_func(input_rref.to_here(), target_rref.to_here())
 
 
 def DistributedLoss(loss: nn.Module, *args: Tuple, **kwargs: Dict) -> Callable:
     loss_func = loss(*args, **kwargs)
 
     def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
-        return rpc.remote(input_rref.owner(), _rloss, args=(loss_func, input_rref, target_rref))
+        return rpc.remote(input_rref.owner(), rloss, args=(loss_func, input_rref, target_rref))
 
     return dloss
 
 
-class MultiProcessPipe(Module):
-    """Wraps an arbitrary :class:`nn.Sequential <torch.nn.Sequential>` module
-    to train on Pipe_. If the module requires lots of memory, Pipe will be
-    very efficient.
-    ::
+class PipelineModule(nn.Module):
+    """Constructs a module on a remote device, possibly at a later time (in case the device is not
+    specified when creating PipelineModule.
+    Args:
+        module_cls (nn.Module): Class for the module to be created remotely.
+        args (Sequence): args to be passed to ``module_cls``.
+        kwargs (Dict, optional): kwargs to be passed to ``module_cls``.
+        num_input (int): number of inputs to the forward function.
+        num_outputs: (int, optional): If the forward function returns a tuple, number of elements
+            in the tuple, otherwise it should be None
+        remote_device: (str, optional): Device on the destination worker where we‘d like to place
+            this module. The format should be "<workername>/<device>", where the device field can be
+            parsed as torch.device type. E.g., "trainer0/cpu", "trainer0", "ps0/cuda:0".
+            If this parameter can be provided later by calling the method instantiate
```

Comment:
is this sentence cut midway? i feel like something should come after instantiate.